### PR TITLE
Export transform and model for torchscript, quantization and deployment

### PR DESCRIPTION
**Pull request by hudeven**
It's troublesome to deal with transform and model separately during torchscript, quantization and deploy with torchserve. So I added `CombinedModule` wrapper to hold them such that it can consume raw input(text) and produce the prediction(generated text) end to end.

### CODE REVIEW

File: torchrecipes/paved_path/README.md [hunk 1]
```diff
@@ -21,23 +22,26 @@ You will get output like below. The snapshot path can be used for inference or r
...
-python charnn/main.py trainer.max_epochs=3 trainer.snapshot_path=/tmp/charnn/run-1f7abaed/snapshots/epoch-1
+torchx run  -s local_cwd dist.ddp -j 1x2 --script main.py
 ```
+> **_NOTE_**: `-j 1x2` specifies single node with 2 GPUs. Learn more about torchx [here](https://pytorch.org/torchx/latest/)
 
-3. Generate text from a model
+* Resume training from a snapshot and train with more epochs
 ```bash
-python charnn/main.py charnn.task="generate" charnn.phrase="hello world" trainer.snapshot_path=/tmp/charnn/run-1f7abaed/snapshots/epoch-1
+python main.py trainer.max_epochs=3 trainer.snapshot_path=/tmp/charnn/my_job/snapshots/last
```
Comment (dracifer) [inline torchrecipes/paved_path/README.md:37]: should we also use `torchx run` here?

File: torchrecipes/paved_path/charnn/combined_module.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        steps: int = 100,
+        temperature: float = 1.0,
+        sample: bool = True,
+        top_k: Optional[int] = 10,
+    ) -> Tensor:
+        """
+        take a conditioning sequence of indices in x (of shape (b,t)) and predict the next token in
+        the sequence, feeding the predictions back into the model each time. Clearly the sampling
+        has quadratic complexity unlike an RNN that is only linear, and has a finite context window
+        of block_size, unlike an RNN that has an infinite context window.
```
Comment (dracifer) [inline torchrecipes/paved_path/charnn/combined_module.py:58]: this is `sample()` function right? we should get rid of it? also, wonder if we have some link to the theory of this logic

File: torchrecipes/paved_path/charnn/export.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    "-t", "--torchscript", help="torchscript the model", action="store_true"
+)
+
+
+def main() -> None:
+    args = parser.parse_args()
+
+    fs, intput_path = fsspec.core.url_to_fs(args.input_path)
+    with fs.open(intput_path, "rb") as f:
+        model = torch.load(f, map_location="cpu")
```
Comment (dracifer) [inline torchrecipes/paved_path/charnn/export.py:38]: shall we use snapshot to save/load trained model as well?

File: torchrecipes/paved_path/charnn/model.py [hunk 4]
```diff
@@ -177,17 +175,13 @@ def _init_weights(self, module) -> None:
...
         assert t <= self.block_size, "Cannot forward, model block size is exhausted."
 
         # forward the GPT model
         x = self.emb_stem(idx)
         x = self.blocks(x)
         x = self.ln_f(x)
         logits = self.head(x)
-        loss = None
-        if targets is not None:
-            loss = F.cross_entropy(logits.view(-1, logits.size(-1)), targets.view(-1))
```
Comment (dracifer) [inline torchrecipes/paved_path/charnn/model.py:191]: make sense to move loss out of the main module!

File: torchrecipes/paved_path/charnn/combined_module.py (hunk 2 above)
Comment (hudeven) [reply torchrecipes/paved_path/charnn/combined_module.py:58]: Right, I think it's originally copied from here: https://github.com/karpathy/minGPT/blob/7218bcfa527c65f164de791099de715b81a95106/mingpt/model.py#L283

I can't find the link to the theory of this logic. As we will publish the tutorial with torchvision example, maybe leave it as it is?

File: torchrecipes/paved_path/charnn/export.py (hunk 3 above)
Comment (hudeven) [reply torchrecipes/paved_path/charnn/export.py:38]: I tried but feel it causes more trouble than benefits. Snapshot save/load state_dicts from training states including model, optimizer, progress, etc. 1) we only need model here. It's a waste to pass the larger snapshot file in S3 2) to load the state_dict, we have to initialize the GPT model object here with the same config as training, which make this script less generic.

File: torchrecipes/paved_path/charnn/model.py (hunk 4 above)
Comment (hudeven) [reply torchrecipes/paved_path/charnn/model.py:191]: This is also required to make model torchscriptable.

File: torchrecipes/paved_path/README.md (hunk 1 above)
Comment (hudeven) [reply torchrecipes/paved_path/README.md:37]: Okay, will add it later.

File: torchrecipes/paved_path/charnn/export.py (hunk 3 above)
Comment (dracifer) [reply torchrecipes/paved_path/charnn/export.py:38]: For "Snapshot save/load state_dicts from training states including model, optimizer, progress, etc.", we should be able to create a separate snapshot that only save model params? but anyway, it sounds reasonable to avoid the overhead. maybe lets check with @yifuwang about the appropriate use case of snapshot

File: torchrecipes/paved_path/charnn/combined_module.py (hunk 2 above)
Comment (dracifer) [reply torchrecipes/paved_path/charnn/combined_module.py:58]: do we plan to remove "sample()"?

File: torchrecipes/paved_path/charnn/char_transform.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+from typing import List
+
+import fsspec
+import torch
+import torch.nn as nn
+from torch import Tensor
+
+
+class CharTransform(nn.Module):
```
[the remark this answers is not in the document]
Comment (hudeven) [reply torchrecipes/paved_path/charnn/char_transform.py:16]: Yes, the dataset is only 1M for this example. We will use iterable datasets/datapipes for the vision tutorial.

File: torchrecipes/paved_path/README.md (hunk 1 above)
Comment (hudeven) [reply torchrecipes/paved_path/README.md:37]: resolved in the latest version in Diffs

File: torchrecipes/paved_path/charnn/combined_module.py (hunk 2 above)
Comment (hudeven) [reply torchrecipes/paved_path/charnn/combined_module.py:58]: yes, resolved in the latest version in Diffs

**Review by dracifer — APPROVED** [review]
